### PR TITLE
fix: Content set bug with cascading attributes

### PR DESCRIPTION
**Pull request by castastrophe**
In a production environment using the umd versions, it was found that not all of the content-set attributes cascaded down to the tab or accordion being rendered.  This PR adds a set of tests to cover each attribute and moves towards hardcoding the properties in the pfe-content-set rather than inheriting them from tabs and accordion and manipulating them on the way in.

- A new alignment observer has been added to translate the `align` from content-set into the `tab-align` on the tabset.
- A cascadeProperties event is fired once the rendered component is finished building.
- A try-catch was wrapped around the history.pushState as this throws a blocking error in iframe settings that prevents the tab navigation from completing.

### Related issues

- (#1398) pfe-content-set | Issues with cascading prefixes in UMD version
- (#1182) pfe-content-set does not work within an iframe < this was accidentally fixed but yay!

### Preview

Link(s) to demo page(s) where this element can be viewed:
- [Demo page](https://deploy-preview-1393--happy-galileo-ea79c4.netlify.app/elements/pfe-content-set/demo/) 

#### Browser requirements

Your component should work in all of the following environments:

- [x] Latest 2 versions of Edge
- [x] Internet Explorer 11 (should be useable, not pixel perfect)
- [x] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [x] Firefox 78 (or latest version for Red Hat Enterprise Linux distribution)
- [x] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [x] Latest 2 versions of Safari
- [x] Android mobile device (such as the Galaxy S9)
- [x] Apple mobile device (such as the iPhone X)
- [x] Apple tablet device (such as the iPhone Pro)

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- [x] Tests have been updated to cover these changes.
- [x] Browser testing passed.
- [x] Repository compiles and tests pass.
- [x] Changelog updated.
- [x] Documentation (README.md, WHY.md, etc.) updated or added.

### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

### CODE REVIEW

**Review by heyMP — APPROVED** [review]
LGTM!

File: elements/pfe-tabs/src/pfe-tabs.js [hunk 1]
```diff
@@ -247,7 +247,12 @@ class PfeTabs extends PFElement {
       const hash = window.location.hash;
 
       urlParams.set(this.id, tab.id);
-      history.pushState({}, "", `${pathname}?${urlParams.toString()}${hash}`);
+      // Attempt to push the history, wrapped in a try/catch for iframe use-case which throws an error
```
[the remark this answers is not in the document]
Comment (castastrophe) [reply elements/pfe-tabs/src/pfe-tabs.js]: No that works great, it's test-fixture + React + Vue

File: elements/pfelement/src/pfelement.js [hunk 2]
```diff
@@ -442,6 +445,8 @@ class PFElement extends HTMLElement {
 
       // If a match was found, cascade each attribute to the element
       if (selectors) {
+        this.log(`Cascade: ${Object.keys(cascade).join(", ")}`);
```
Comment (castastrophe) [inline elements/pfelement/src/pfelement.js:448]:
```suggestion
```

File: elements/pfelement/src/pfelement.js [hunk 3]
```diff
@@ -442,6 +445,8 @@ class PFElement extends HTMLElement {
...
+        this.log(`Selectors: ${selectors.join(", ")}`);
```
Comment (castastrophe) [inline elements/pfelement/src/pfelement.js:449]:
```suggestion
```

File: elements/pfelement/src/pfelement.js [hunk 4]
```diff
@@ -519,6 +524,7 @@ class PFElement extends HTMLElement {
     for (let mutation of mutationsList) {
       // If a new node is added, attempt to cascade attributes to it
       if (mutation.type === "childList" && mutation.addedNodes.length) {
+        this.log(`Parse observer firing cascade?`);
```
Comment (castastrophe) [inline elements/pfelement/src/pfelement.js:527]:
```suggestion
```

File: elements/pfelement/src/pfelement.js [hunk 5]
```diff
@@ -322,19 +322,8 @@ class PFElement extends HTMLElement {
     if (window.ShadyCSS) window.ShadyCSS.styleElement(this);
 
     // If the slot definition exists, set up an observer
-    if (typeof this.slots === "object") {
+    if (typeof this.slots === "object")
```
Comment (castastrophe) [inline elements/pfelement/src/pfelement.js:325]: @kylebuch8 Can you sanity check me on this?  I was debugging looping issues in IE11 and I believe it makes sense to move the slot observer into the rendering method.  What do you think?

File: elements/pfelement/src/pfelement.js [hunk 6]
```diff
@@ -398,6 +387,20 @@ class PFElement extends HTMLElement {
...
     this.resetContext();
 
+    // If the slot definition exists, set up an observer
+    if (typeof this.slots === "object" && this._slotsObserver) {
+      this._slotsObserver.observe(this, { childList: true });
+      this._initializeSlots(this.tag, this.slots);
+    }
+
+    // If an observer was defined, set it to begin observing here
+    if (this._cascadeObserver)
```
Comment (castastrophe) [inline elements/pfelement/src/pfelement.js:397]: @kylebuch8 Same question around the cascade observer.  Maybe it makes sense to wait to attach the cascade observer until the component upgrades?

File: elements/pfe-content-set/test/pfe-content-set_test.js [hunk 7]
```diff
@@ -1,146 +1,150 @@
+import { create } from "browser-sync";
```
Comment (castastrophe) [inline elements/pfe-content-set/test/pfe-content-set_test.js:1]:
```suggestion
```

File: elements/pfelement/src/pfelement.js (hunk 6 above)
Comment (heyMP) [reply elements/pfelement/src/pfelement.js:398]: I'm thinking about scenarios where the element is re-running the render() method when properties change.  Would it be duplicating the observer?

Comment (castastrophe) [reply elements/pfelement/src/pfelement.js:398]: Shoot yeah I think you're right.  What's a better place to include it or is there a way to query for it's existence first?

File: elements/pfe-content-set/src/pfe-content-set.html [hunk 8]
```diff
@@ -1,2 +1,2 @@
-<slot id="lightdom" hidden></slot>
+<div hidden><slot id="lightdom"></slot></div>
```
Comment (heyMP) [inline elements/pfe-content-set/src/pfe-content-set.html:1]: Nice! I ran into this with pfe-clipboard.

File: elements/pfelement/src/pfelement.js (hunk 6 above)
Comment (heyMP) [reply elements/pfelement/src/pfelement.js:398]: So this is interesting.  The `observe` method is smart enough to remove and add and remove observers on its own.  

> If you call observe() on a node that's already being observed by the same MutationObserver, all existing observers are automatically removed from all targets being observed before the new observer is activated.

https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/observe#reusing_mutationobservers

Comment (castastrophe) [reply elements/pfelement/src/pfelement.js:398]: Nice! Awesome find. I was seeing these observers from pfelement repeatedly run in content set during the updating of the rendering.

Comment (heyMP) [reply elements/pfelement/src/pfelement.js:398]: Did some superficial performance testing and didn't see any glaring differences between moving this into the render method and keeping it in connectedCallback.   I wonder if we could get @zhawkins on the case.

**Review by kylebuch8 — APPROVED** [review]
Lava Glowing Touch Melts 🥵

**Pull request merged**